flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java

```diff
@@ -60,138 +68,153 @@
-                return;
+                return rel;
             }
             CatalogTable catalogTable = tableSourceTable.contextResolvedTable().getResolvedTable();
             List<String> partitionKeys = catalogTable.getPartitionKeys();
             if (partitionKeys.isEmpty()) {
-                factSideFactors.isSuitableFactScanSource = false;
-                return;
+                return rel;
             }
             DynamicTableSource tableSource = tableSourceTable.tableSource();
             if (!(tableSource instanceof SupportsDynamicFiltering)
                     || !(tableSource instanceof ScanTableSource)) {
-                factSideFactors.isSuitableFactScanSource = false;
-                return;
+                return rel;
             }
             if (!isNewSource((ScanTableSource) tableSource)) {
-                factSideFactors.isSuitableFactScanSource = false;
-                return;
+                return rel;
             }
 
             List<String> candidateFields =
                     joinKeys.stream()
                             .map(i -> scan.getRowType().getFieldNames().get(i))
                             .collect(Collectors.toList());
             if (candidateFields.isEmpty()) {
-                factSideFactors.isSuitableFactScanSource = false;
-                return;
+                return rel;
             }
 
-            factSideFactors.isSuitableFactScanSource =
-                    !getSuitableDynamicFilteringFieldsInFactSide(tableSource, candidateFields)
-                            .isEmpty();
-        } else if (rel instanceof HepRelVertex) {
-            visitFactSide(((HepRelVertex) rel).getCurrentRel(), factSideFactors, joinKeys);
+            List<String> acceptedFilterFields =
+                    getSuitableDynamicFilteringFieldsInFactSide(tableSource, candidateFields);
+
+            if (acceptedFilterFields.size() == 0) {
+                return rel;
+            }
+
+            // Apply suitable accepted filter fields to source.
+            ((SupportsDynamicFiltering) tableSource).applyDynamicFiltering(acceptedFilterFields);
+
+            List<Integer> acceptedFieldIndices =
+                    acceptedFilterFields.stream()
+                            .map(f -> scan.getRowType().getFieldNames().indexOf(f))
+                            .collect(Collectors.toList());
+            List<Integer> dynamicFilteringFieldIndices = new ArrayList<>();
+            for (int i = 0; i < joinKeys.size(); ++i) {
+                if (acceptedFieldIndices.contains(joinKeys.get(i))) {
+                    dynamicFilteringFieldIndices.add(dimSideJoinKey.get(i));
+                }
+            }
+
+            BatchPhysicalDynamicFilteringDataCollector dynamicFilteringDataCollector =
+                    createDynamicFilteringConnector(dimSide, dynamicFilteringFieldIndices);
+
+            factSideFactors.isChanged = true;
+
+            if (!isSuitableJoin(join)) {
```

Comment:
this validation can be moved to `join` branch

Reply:
> this validation can be moved to `join` branch

Done!